contrib/clients/xmail/xmail.man
.\"
.\" @(#)xmail.l 1.4 91/09/27 NSC;
.\"
.\" Copyright 1990,1991 by National Semiconductor Corporation.
.\"
.\" Permission to use, copy, modify, and distribute this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation, and that the name of National Semiconductor Corporation not
.\" be used in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.  National Semiconductor
.\" Corporation makes no representations about the suitability of this software
.\" for any purpose. It is provided "as is" without express or implied warranty.
.\"
.\" NATIONAL SEMICONDUCTOR CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
.\" FITNESS, IN NO EVENT SHALL NATIONAL SEMICONDUCTOR CORPORATION BE LIABLE FOR
.\" ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Author:  Michael C. Wagnitz - National Semiconductor Corporation
.\"
.ds nS National Semiconductor Corporation
.ds xM xmail
.if n .ds Q \&"
.if n .ds U \&"
.if t .ds Q ``
.if t .ds U ''
..
.\"
.TH XMAIL 1 "September 27, 1991" "NSC"
.SH NAME
xmail \- X11 visual interface to the mail program
.SH SYNOPSIS
.B \*(xM
.RB "[\|" \-\fItoolkitoptions\fP "\|]"
.RB "[\|" \-\fIxmailoptions\fP "\|]"
.SH DESCRIPTION
.B \*(xM
is an X11 window based visual interface to the
.B Mail
program.  It offers functionality similar to the SunView mailtool, but with
some differences. 

Users may use mouse button actions in the index window to select a message for
further processing (by pressing the left mouse button), or automatically read
the selected message (by pressing the right mouse button).

Folder specification is aided by a selection menu which can walk down through
the user's folder directory hierarchy. 

Message composition and modification is provided by the user's preferred
editor (/usr/ucb/vi by default).  Editor preference may be specified using
either the user's process or Mail environment variable
.SB VISUAL, 
or the xmail \fBeditorCommand\fR command line or application resource.

A short description of the command button action is presented in the
status window whenever the cursor enters any of the xmail command button
windows.  Additional help information is presented for any command button or
other part of xmail, by pressing the middle mouse button on the window or
command button while holding down either Shift key.

Support is also provided for displaying the content of an X-Face mail header,
and for adding your own X-Face header to outgoing mail if present.
.SH OPTIONS
An unrecognized or illegal command line option specification will cause
display of a usage statement and termination of \*(xM.
.TP
\fB\-toolkitoptions\fR
All standard X11 toolkit options are accepted on the command line.
.TP
\fB\-xmailoptions\fR
\*(xM also accepts the following additional options:
.RS
.TP
\fB\-buttonfont \fIfontname\fR
specifies the font to use when creating command buttons.  This option overrides
the \*(xM resource specification \*Q\fB*buttonFont: \fIfontname\fR\*U.
.TP
\fB\-e\fR
Sets the \*(xM resource specification \*Q\fB*expert: \fITrue\fR\*U.  This
resource determines whether \*Q\fBconfirm\fR\*U popups will be presented
when committing changes to a folder or deleting a new message composition.
By default the resource is False, which means confirm popups will be used.
.TP
\fB\-editorCommand \fI<quoted command string>\fR
Specifies the editor resource that \*(xM should use when editing a message
composition.  The quoted command must be an \fIsprintf\fR\-able format string
containing two string declarations.  The first of these declarations will be
replaced with the name of the display connection for the xmail process, to
allow the edit window to appear on the correct display screen.  The second
declaration will be replaced with the name of the file to be edited.
If \*(xM cannot locate the two separate string declarations in the command,
the command will be rejected and processing will default to the older
.SM VISUAL
methodology.
.TP
\fB\-f\ \fI+foldername\fR
tells Mail to start by processing the specified folder.
This is also a Mail program option, passed to it during startup.
The option can also be specified as an \*(xM resource
using the specification \*Q\fB*MFileName: \fI+foldername\fR\*U.
.TP
\fB\-fn\ \fI+fontname\fR
tells \*(xM to use the indicated font for all windows and buttons.
.TP
\fB\-font\ \fI+fontname\fR
tells \*(xM to use the indicated font for all windows and buttons.
.TP
\fB\-h\fR
tells \*(xM to include the host name in the icon window when displaying
the icon.  The option can also be specified as an \*(xM resource
using the specification \*Q\fB*icon*useHost: \fITrue\fR\*U.
.TP
\fB\-helpfont \fIfontname\fR
specifies the font to use when displaying help text.  This option overrides
the \*(xM resource specification \*Q\fB*helpFont: \fIfontname\fR\*U.
.TP
\fB-iconGeometry \fI[+-]XOff[+-]YOff\fR
provides a command line option for specifying the preferred geometry
location for the \*(xM icon window.  This overrides any \*(xM resource
specification for \*Q\fB*iconGeometry: \fI[+-]XOff[+-]YOff\fR\*U.
.TP
\fB\-ls\fR
tells \*(xM to \fBnot\fR show the last message of a folder by default.  This
option is equivalent to setting the \*(xM resource \fB*Show_Last\fR to
\fIFalse\fR.  If disabled, \*(xM (actually Mail) will display the default
(usually first) message of a folder when switching to or re-reading a folder.
By default, \*(xM attempts to read the latest (or last) message of a folder.
.TP
\fB\-n\fR
tells Mail to \fBnot\fR initialize from the system default Mail.rc file.
This option is actually a Mail program option, which is passed on to it during
startup from \*(xM.
The option can also be specified as an \*(xM resource
using the specification \*Q\fB*mailopt_n: \fITrue\fR\*U.
.TP
\fB\-nb\fR
tells \*(xM to \fBnot\fR ring the terminal bell when issuing status and
error messages.  This is equivalent to setting the \*(xM resource
specification \*Q\fB*bellRing: \fIFalse\fR\*U.  The default will be to
ring the bell for error and appropriate status messages.
.TP
\fB\-noinfo\fR
tells \*(xM to \fBnot\fR display info messages when entering or leaving a
command button window with the mouse cursor.  Normally, the enter/leave
window events will change the text displayed in the status window, to show
information about the command button currently under the mouse cursor.  This
option allows you to disable this new feature, if desired.
This is equivalent to setting the \*(xM resource specification
\*Q\fB*show_Info: \fIFalse\fR\*U.
.TP
\fB\-nx\fR
tells \*(xM to \fBnot\fR include any X-Face header strings when displaying the
text of an incoming message.  This is equivalent to setting the \*(xM resource
specification \*Q\fB*No_X_Hdr: \fITrue\fR\*U.  The default will be to include
all text of the incoming message, including any X-Face header strings.  Either
setting will not deter the display of the X-Face bitmap if it exists in the
message headers (and the X_FACE option was specified during compilation).
.TP
\fB\-rv\fR
tells \*(xM to reverse the foreground and background colors of the icon
window when displaying the icon.  The option can also be specified as an
\*(xM resource using the specification \*Q\fB*icon*reverseVideo: \fITrue\fR\*U.
.TP
\fB\-textfont \fIfontname\fR
specifies the font to use when creating text windows.  This option overrides
the \*(xM resource specification \*Q\fB*textFont: \fIfontname\fR\*U.
.TP
\fB\-u\fR
tells \*(xM to include the user login name in the icon window when displaying
the icon.  The option can also be specified as an \*(xM resource
using the specification \*Q\fB*icon*useName: \fITrue\fR\*U.
.TP
\fB\-U\fR
tells Mail to convert internet style addresses into uucp format.
This option also is a Mail option, passed on to it during startup.
The option can also be specified as an \*(xM resource
using the specification \*Q\fB*mailopt_U: \fITrue\fR\*U.
.RE
.SH USAGE
The \*(xM shell appearance is similar in layout to mailtool.
A title bar decorates the top of the shell, and carries the name and
version of the program, as well as information about the current mail folder
and its content.

Immediately below the title bar, an index window displays header information
for the current mail folder.
Message selection is indicated by a \fBgreater-than '>'\fR symbol located at
the beginning of the header line for the particular message.
New and unread messages are indicated by '\fBN\fR' and '\fBU\fR' characters
respectively, preceding the message number.
Deleted messages not yet removed from the index list have the letter '\fBD\fR'
preceding the message number.
Saved messages are now marked with the letter '\fBS\fR', to denote that the
message would normally disappear from your system folder when you finally
commit the changes.
Messages that are marked for preservation in the system folder are now marked
with the letter '\fBP\fR' until the folder changes are committed.
Keyboard keys can be used to scroll the index window,
in addition to the more traditional mouse on scrollbar mechanism.

Below the index
window, the status window displays information, status, and error messages.
As the cursor is moved between the various command button windows in the xmail
program, a short description of the particular button's action(s) is displayed
in the status window.  Leaving the command button window restores the default
information about obtaining help on any part of xmail.
Error and status messages are also displayed in this same status window.
The terminal bell is usually rung to draw attention to these messages.
The status bell can be disabled from the command line or user resource file.

Immediately below the status window and attached to it, the xmail
shell displays a panel with a selection of commonly used Mail commands.
Pressing and then releasing (\fIalso known as \*Q\fBclicking\fP\*U)\fR
the \fBleft\fR mouse button while the cursor is within a
particular command button box invokes that Mail command.

A grip at the top right corner of the command panel permits the user to slide
the status window/command panel combination up and down in the \*(xM shell, to
accommodate those times when the user is reviewing a short list of longer than
average messages, or a longer list of relatively short messages.

Each button in the command panel also offers an expanded menu of choices.
Pressing the \fBright\fR mouse button presents the user with the
expanded menu.  \fBDragging\fR the cursor over one of these selections
\fIwhile still holding down the right button\fR causes the particular
selection to highlight.  A description of the highlighted command's action is
also displayed in the status window.  \fBReleasing\fR the right button while a
selection is highlighted invokes that particular command action.

Keyboard accelerators for these actions are indicated by appropriate entries
in the command menus.  Users may take advantage of these accelerators by
holding down the Shift and/or Control key(s) prior to clicking the left mouse
button on the desired command, in lieu of raising the command popup menu.

Message text is displayed in the bottom window of the \*(xM shell.
Like the index window above, scroll bars provide easy access to any portion of
the text.  Keyboard keys can also be used to scroll the text window.  If the
displayed message contains an X-Face header string, the content of that string
is displayed as a bitmap in the upper right-hand corner of the text window.
(\fIThis feature is a compile time option, and requires the existence of the
\fBcompface\fP library routines, which are not supplied with the xmail sources.
See the section on \s-1X-FACE HEADERS\s+1 for details on acquiring
this library.\fR)

In addition, both the text window and index window have duplicate keyboard
definitions to supplant using the
mouse on the command buttons for normal \*(xM command operation.  Like many of
the resources for xmail, these key bindings are now declared in the application
resource file, to allow the use to modify them as desired.


contrib/clients/xmail/xmail.c
/*
 * xmail - X window system interface to the mail program
 *
 * Copyright 1989 The University of Texas at Austin
 *
 * Author:	Po Cheung
 * Date:	March 10, 1989
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.  The University of Texas at Austin makes no 
 * representations about the suitability of this software for any purpose.  
 * It is provided "as is" without express or implied warranty.
 *
 * Copyright 1990,1991 by National Semiconductor Corporation
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of National Semiconductor Corporation not
 * be used in advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 *
 * NATIONAL SEMICONDUCTOR CORPORATION MAKES NO REPRESENTATIONS ABOUT THE
 * SUITABILITY OF THIS SOFTWARE FOR ANY PURPOSE.  IT IS PROVIDED "AS IS"
 * WITHOUT EXPRESS OR IMPLIED WARRANTY.  NATIONAL SEMICONDUCTOR CORPORATION
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  IN NO
 * EVENT SHALL NATIONAL SEMICONDUCTOR CORPORATION BE LIABLE FOR ANY SPECIAL,
 * INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
 * LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
 * OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 *
 * The following software modules were created and are Copyrighted by National
 * Semiconductor Corporation:
 *
 * 1. initialize: and
 * 2. SetHints.
 *
 * Author:  Michael C. Wagnitz - National Semiconductor Corporation
 *
 */

#include "global.h"
#include "Mailwatch.h"			/* use as icon and watch for newmail */
#include <X11/Xaw/CommandP.h>
#include <X11/Xaw/TextP.h>
#include <X11/cursorfont.h>		/* use watch cursor for busy notify */
#include <X11/bitmaps/cross_weave>	/* background use in Newmail notify */

#ifndef	lint
static char what[] =
 "@(#)xmail.c 1.4 91/09/20 Copyright 1989,1990,1991 National Semiconductor Corp.";
#endif

#ifndef	DEFAULT_PROMPT
#define	DEFAULT_PROMPT	"& "
#endif

#define Offset(field) (XtOffset(XmailResources *, field))

static XtResource resrcs[] = {
    {"textFont", XtCFont, XtRFontStruct, sizeof (XFontStruct *),
	Offset(textFont), XtRString, XtDefaultFont},
    {"helpFont", XtCFont, XtRFontStruct, sizeof (XFontStruct *),
	Offset(helpFont), XtRString, XtDefaultFont},
    {"buttonFont", XtCFont, XtRFontStruct, sizeof (XFontStruct *),
	Offset(buttonFont), XtRString, XtDefaultFont},
    {"iconGeometry", "IconGeometry", XtRString, sizeof(char *),
        Offset(iconGeometry), XtRString, (caddr_t) NULL},
    {"mFileName", "MFileName", XtRString, sizeof(char *), 
	Offset(MFileName), XtRString, (caddr_t) NULL},
    {"editorCommand", "EditorCommand", XtRString, sizeof(char *), 
	Offset(editorCommand), XtRString, (caddr_t) NULL},
    {"bellRing", "BellRing", XtRBoolean, sizeof(Boolean), 
	Offset(bellRing), XtRImmediate, (caddr_t) True},
    {"expert", "Expert", XtRBoolean, sizeof(Boolean), 
	Offset(expert), XtRImmediate, (caddr_t) False},
    {"iconic", "Iconic", XtRBoolean, sizeof(Boolean), 
	Offset(iconic), XtRImmediate, (caddr_t) False},
    {"mailopt_n", "Mailopt_n", XtRBoolean, sizeof(Boolean), 
	Offset(mailopt_n), XtRImmediate, (caddr_t) False},
    {"mailopt_U", "Mailopt_U", XtRBoolean, sizeof(Boolean), 
	Offset(mailopt_U), XtRImmediate, (caddr_t) False},
    {"show_Last", "Show_Last", XtRBoolean, sizeof(Boolean), 
	Offset(Show_Last), XtRImmediate, (caddr_t) True},
    {"show_Info", "Show_Info", XtRBoolean, sizeof(Boolean), 
	Offset(Show_Info), XtRImmediate, (caddr_t) True},
    {"no_X_Hdr", "No_X_Hdr", XtRBoolean, sizeof(Boolean), 
	Offset(No_X_Hdr), XtRImmediate, (caddr_t) False},
    { XtNborderWidth, XtCBorderWidth, XtRInt, sizeof (int),
	Offset(borderWidth), XtRString, "1"},
};

#undef Offset

static XrmOptionDescRec Opts[] = {
    {"-borderwidth",	".TopLevelShell.borderWidth",	XrmoptionSepArg,	(caddr_t) NULL},
    {"-buttonfont",	"*buttonFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-buttonFont",	"*buttonFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-bw",		".TopLevelShell.borderWidth",	XrmoptionSepArg,	(caddr_t) NULL},
    {"-editorcommand",	"*editorCommand",		XrmoptionSepArg,	(caddr_t) NULL},
    {"-editorCommand",	"*editorCommand",		XrmoptionSepArg,	(caddr_t) NULL},
    {"-e",		"*expert",			XrmoptionNoArg,		(caddr_t) "True"},
    {"-f",		"*MFileName",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-fn",		"*Font",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-font",		"*Font",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-geometry",	".geometry",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-helpfont",	"*helpFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-helpFont",	"*helpFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-iconGeometry",	"*iconGeometry",		XrmoptionSepArg,	(caddr_t) NULL},
    {"-iconic",		"*iconic",			XrmoptionNoArg,		(caddr_t) "True"},
    {"-ls",		"*show_Last",			XrmoptionNoArg,		(caddr_t) "False"},
    {"-m",		"*icon*useHost",		XrmoptionNoArg,		(caddr_t) "True"},
    {"-n",		"*mailopt_n",			XrmoptionNoArg,		(caddr_t) "True"},
    {"-nb",		"*bellRing",			XrmoptionNoArg,		(caddr_t) "False"},
    {"-noinfo",		"*show_Info",			XrmoptionNoArg,		(caddr_t) "False"},
    {"-nx",		"*no_X_Hdr",			XrmoptionNoArg,		(caddr_t) "True"},
    {"-rv",		"*icon*reverseVideo",		XrmoptionNoArg,		(caddr_t) "True"},
    {"-textfont",	"*textFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-textFont",	"*textFont",			XrmoptionSepArg,	(caddr_t) NULL},
    {"-U",		"*mailopt_U",			XrmoptionNoArg,		(caddr_t) "True"},
    {"-u",		"*icon*useName",		XrmoptionNoArg,		(caddr_t) "True"},
    {"-xrm",		NULL,				XrmoptionResArg,	(caddr_t) NULL},
};

Atom		wmDeleteWindow;

int		mailargc;		/* counter passed to mail child	*/
int		RootWidth, RootHeight;
int		Highlighted;		/* state of 'Newmail' highlighting */

char		Command[BUFSIZ];	/* xmail command string */
char	 	InReply[BUFSIZ];	/* reply reference string */
char	 	tmpName[BUFSIZ];	/* message temporary filename */
char		*mailargv[7];		/* array passed to mail child */

Pixmap		hatch;			/* cross_weave used to note Newmail */
Widget		toplevel; 		/* top level shell widget */
Window		WaitCursorWindow;
helpList	HelpList;

XmailResources 	XMail;			/* application resources of xmail */

/*
** @(#)mailoptions() - construct command line arguments for calling mail.
**                     Return the argument list for mail and new value of argc.
*/
int
mailoptions()
{
 XWMHints	*wm_hints;
 int		i;
 char		*Mailpgm;		/* name of executable Mailpgm */


 if (! (Mailpgm = (char *)getenv("XMAILER")))	/* first looks up env var */
    Mailpgm = DEFAULT_MAILER;


 wm_hints = XGetWMHints(XtDisplay(toplevel), XtWindow(toplevel));

 i = 0;
 mailargv[i++] = Mailpgm;	/* Argv[0] is the name of the program */
 mailargv[i++] = "-N";		/* no version or header info at start */
 if (XMail.mailopt_n)
    mailargv[i++] = "-n";	/* don't initialize from Mail.rc file */
 if (XMail.mailopt_U)
    mailargv[i++] = "-U";	/* Change uucp to Internet addresses */

 if (wm_hints->initial_state == IconicState) {	/* start iconic in bogus */
    FILE	*fp;
    char	bogus[BUFSIZ];

    (void) sprintf(bogus, "%s+", tmpName);
    if (fp = fopen(bogus, "w")) {	/* create our bogus mail file */
       (void) fprintf(fp, "\n");
       (void) fclose(fp);
       mailargv[i++] = "-f";		/* start in our bogus mail folder */
       mailargv[i++] = bogus;
       In_Bogus_Mail_File = True;
      }
   } else if (XMail.MFileName) {
	     mailargv[i++] = "-f";	/* start from mail folder MFileName */
	     mailargv[i++] = XMail.MFileName;
            }
 mailargv[i] = NULL;
 XFree(wm_hints);
 return (i);
} /* mailoptions */


/*
** @(#)initialize() - establish xmail program defaults and setups
*/
void
initialize()
{
 String		disp, file;
 long		buttonSpace, i;
 Display	*dpy = XtDisplay(toplevel);
 char		*str_type[20];
 XrmValue	value;


 if (XrmGetResource(dpy->db, "command.borderWidth", "Command.BorderWidth",
    str_type, &value)) {
    if ((i = atoi(value.addr)) > XMail.borderWidth)
       XMail.borderWidth = i;
   }

 if (XrmGetResource(dpy->db, "send.borderWidth", "Send.BorderWidth",
    str_type, &value)) {
    if ((i = atoi(value.addr)) > XMail.borderWidth)
       XMail.borderWidth = i;
   }

 HelpList.indx = 0;

 (void) strcpy(Command, "Start");	/* let parser know we've started */

 (void) sprintf(tmpName, "/tmp/xmail%d", getpid());
/*
** TRAP any stupid attempt to set the border width of this application to less
** than the allowed minimum (0) or greater than an acceptable maximum.  This
** to prevent a really wierd reaction to a negative number for border width.
*/
 if (XMail.borderWidth < 0)
     XMail.borderWidth = 1;
 if (XMail.borderWidth > 6)
     XMail.borderWidth = 6;
/*
** Compute all window dimensions based on extents of the specified fonts.
** Make shell wide enough to hold eight buttons on a row.  Total button width
** includes width of text area plus width of separation plus width of border.
*/
 XMail.commandHSpace	= 10;
 XMail.commandVSpace	= 10;
 XMail.buttonHeight	= CHARHEIGHT(XMail.buttonFont) + (XMail.buttonFont->ascent / 2);
 XMail.buttonWidth	= figureWidth(XMail.buttonFont) * 9; /* ' NewMail ' */
       buttonSpace	= XMail.borderWidth * 2;
       buttonSpace	+= XMail.buttonWidth + XMail.commandHSpace;
 XMail.shellWidth	= (8 * buttonSpace) + 26;    /*  8 buttons per row */
 XMail.fileBoxWidth	= (4 * buttonSpace) + XMail.buttonWidth;
 XMail.helpHeight	= CHARHEIGHT(XMail.helpFont) * 14 + 5;
 XMail.helpWidth	= figureWidth(XMail.helpFont) * 60 + 5;
 XMail.helpX		= (XMail.shellWidth - XMail.helpWidth) / 2;
 XMail.helpY		= 70;
       buttonSpace	= CHARHEIGHT(XMail.textFont) + (XMail.textFont->ascent / 2);
 XMail.indexHeight	= 12 * buttonSpace;
 XMail.textHeight	= 23 * buttonSpace;
 XMail.commandHeight	= (XMail.buttonHeight*2) + (XMail.commandVSpace*3) +
			  (4 * XMail.borderWidth);
 XMail.menuX		= 15;
 XMail.menuY		= 7;
       RootWidth	= XDisplayWidth(XtDisplay(toplevel),
			  DefaultScreen(XtDisplay(toplevel))) - 1;
       RootHeight	= XDisplayHeight(XtDisplay(toplevel),
			   DefaultScreen(XtDisplay(toplevel))) - 1;
/*
** Ensure xmail shell height does not exceed the height of the root window.
*/
      buttonSpace	= (2 * XMail.buttonHeight) + XMail.indexHeight +
			  XMail.commandHeight + XMail.textHeight + 
			  (6 * XMail.borderWidth);

 for (i = 0, buttonSpace -= RootHeight; buttonSpace > 0; i++) {
     buttonSpace -= CHARHEIGHT(XMail.textFont);
     if (i % 2)
        XMail.indexHeight -= CHARHEIGHT(XMail.textFont);
     else
        XMail.textHeight -= CHARHEIGHT(XMail.textFont);
    }
/*
** If editorCommand resource exists, warn if it doesn't have two %s entries.
*/
 if (XMail.editorCommand)
    if (! (disp =  strchr(XMail.editorCommand, '%')) || *(disp + 1) != 's' ||
        ! (file = strrchr(XMail.editorCommand, '%')) || *(file + 1) != 's' ||
           disp == file) {
    XtWarning("xmail editorCommand resource improperly defined... ignoring");
    XtFree((char *) XMail.editorCommand);
    XMail.editorCommand = NULL;
   }

 parser_init();
} /* initialize */


/*
** @(#)SetHints() - Sets hatching, cursor, icon and size hints, and wm protocol
*/
SetHints()
{
 int			x, y;
 unsigned long		valuemask;
 CommandWidget		nm = (CommandWidget)XtNameToWidget(toplevel, "topBox.commandPanel.Newmail");
 Display		*dpy = XtDisplay(toplevel);
 Screen			*scn = XtScreen(toplevel);
 Window			win = XtWindow(toplevel);
 XSetWindowAttributes	attributes;
 XSizeHints		size_hints;
 XWMHints		wm_hints;
/*
** Hatch will be used to indicate new mail arrival on the command button
*/
 hatch	= XCreatePixmapFromBitmapData(dpy, XtScreen(toplevel)->root,
			cross_weave_bits, cross_weave_width, cross_weave_height,
			nm->label.foreground, nm->core.background_pixel,
			DefaultDepth(dpy, DefaultScreen(dpy)));
/*
** Notify the window manager about our icon window
*/
 wm_hints.input		= True;
 wm_hints.initial_state	= (XMail.iconic) ? IconicState : NormalState;
 wm_hints.flags		= InputHint | StateHint;

 if (XMail.iconGeometry) {
    ParseIconGeometry(XMail.iconGeometry, &x, &y);
    wm_hints.icon_x	= x;
    wm_hints.icon_y	= y;
    wm_hints.flags	|= IconPositionHint;
   }
 wm_hints.flags		|= IconWindowHint;
 wm_hints.icon_window	= XtWindow(XtNameToWidget(toplevel, "icon"));

 XSetWMHints(dpy, win, &wm_hints);
/*
** Set starting position and default geometry
*/
 if (! XGetNormalHints(dpy, win, &size_hints)) {
    size_hints.x = 56;
    size_hints.y = 56;			/* slightly off from top left corner */
   }

 size_hints.width	= XMail.shellWidth;
 size_hints.height	= (2 * XMail.buttonHeight) + XMail.indexHeight +
			  XMail.commandHeight + XMail.textHeight +
			  (6 * XMail.borderWidth);

 size_hints.min_width = size_hints.max_width = size_hints.width;
 size_hints.min_height = size_hints.max_height = size_hints.height;
 size_hints.flags = USPosition | USSize | PMinSize;

 XSetNormalHints(dpy, win, &size_hints);
/*
** Add a protocol flag indicating we wish to handle WM_DELETE_WINDOW requests
*/
 wmDeleteWindow = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
 XSetWMProtocols(dpy, win, &wmDeleteWindow, 1);
/*
** Create a WaitCursor window, ala FAQ to Ignore device events while
** the busy cursor is displayed.  The window will be as big as the
** display screen, and clipped by its own parent window, so we never
** have to worry about resizing.
*/
 valuemask = CWDontPropagate | CWCursor;

 attributes.do_not_propagate_mask =  (KeyPressMask | KeyReleaseMask |
                                   ButtonPressMask | ButtonReleaseMask |
                                   PointerMotionMask);

 attributes.cursor = XCreateFontCursor(dpy, XC_watch);

 WaitCursorWindow = XCreateWindow(dpy, win, 0, 0, WidthOfScreen(scn),
                          HeightOfScreen(scn), (unsigned int) 0, CopyFromParent,
                          InputOnly, CopyFromParent, valuemask, &attributes);
} /* SetHints */


/*
** @(#)main() - main routine for the x11 interface to the mail utility
*/
main(argc, argv)
int	argc;
char	**argv;
{
 char	*cp, **aV;
 int	indx, value, aC;

 static XtActionsRec actionTable[] = {
	{"CheckInsert",		(XtActionProc) CheckInsert},
	{"DeleteChar",		(XtActionProc) DeleteChar},
	{"DeleteLine",		(XtActionProc) DeleteLine},
	{"DeleteWord",		(XtActionProc) DeleteWord},
	{"DoCmd",		(XtActionProc) DoCmd},
	{"DoNothing",		(XtActionProc) DoNothing},
	{"DoReply",		(XtActionProc) DoReply},
	{"DoSave",		(XtActionProc) DoSave},
	{"DoSelected",		(XtActionProc) DoSelected},
	{"Iconify",		(XtActionProc) Iconify},
	{"Folder",		(XtActionProc) Folder},
	{"MyNotify",		(XtActionProc) MyNotify},
	{"NextField",		(XtActionProc) NextField},
	{"PrintMsg",		(XtActionProc) PrintMsg},
	{"Quit",		(XtActionProc) Quit},
	{"SetAliases",		(XtActionProc) SetAliases},
	{"SetDirectory",	(XtActionProc) SetDirectory},
	{"SetFolders",		(XtActionProc) SetFolders},
	{"SetMenu",		(XtActionProc) SetMenu},
	{"SetPopup",		(XtActionProc) SetPopup},
	{"SetSelect",		(XtActionProc) SetSelect},
	{"ShowHelp",		(XtActionProc) ShowHelp},
        {NULL, NULL}
      };
/*
** First, examine our command line arguments and Class resources for bogus
** values which would cause the application to operate incorrectly.  Drop
** command lines arguments and/or change the resource values before we
** initialize our widgets.
*/
 aV = (char **) XtMalloc((unsigned) (argc + 1) * sizeof(char *));
 aC = 0;

 for (value = indx = 0; indx < argc; indx++)
      value += strlen(argv[indx]) + 1;
 aV[aC] = (char *) XtMalloc((unsigned) value);

 (void) strcpy(aV[aC++], *argv);
 aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
 for (indx = 1; indx < argc; indx++)
     switch (argv[indx][0]) {		/* look for, -xrm *[Bb]orderWidth or */
         case '-':			/* -bw, w/ negative valued arguments */
              if (argv[indx][1] != 'b' && argv[indx][2] != 'w' &&
                  argv[indx][1] != 'x' && argv[indx][2] != 'r' &&
                  argv[indx][3] != 'm') {
                 (void) strcpy(aV[aC++], argv[indx]);
                 aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                }
              else if (sscanf(argv[indx+1], "%d", &value) == 1) {
                   if (value < 0)
                      indx++;		/* skip the value argument too */
                   else if (value < 7) {
                           (void) strcpy(aV[aC++], argv[indx]);
                           aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                          } else {
                           (void) strcpy(aV[aC++], argv[indx++]);
                           aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                           (void) strcpy(aV[aC++], "6");
                           aV[aC] = aV[aC - 1] + 2;
                          }
                }
              else if ((cp = strchr(argv[indx + 1], 'o')) == NULL ||
                       strncmp(cp, "orderWidth:", 11)     != 0    ||
                       sscanf(cp + 11, "%d", &value)      != 1) {
                      (void) strcpy(aV[aC++], argv[indx]);
                      aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                     }
              else if (value >= 0) {
                      if (value < 7) {
                         (void) strcpy(aV[aC++], argv[indx]);
                         aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                        } else {
                         (void) strcpy(aV[aC++], argv[indx++]);
                         aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
                         (void) strcpy(aV[aC], argv[indx]);
                         cp = strchr(aV[aC], ':');
                         *++cp = '\0';
                         (void) strcat(aV[aC], " 6");	/* max allowed value */
                         aV[aC + 1] = aV[aC] + strlen(aV[aC]) + 1;
                         aC++;
                        }
                }
              else indx++;		/* skip the value argument too */
              break;			/* if negative, it is not included */
         default :
              (void) strcpy(aV[aC++], argv[indx]);
              aV[aC] = aV[aC - 1] + strlen(aV[aC - 1]) + 1;
              break;
        }
 aV[aC] = NULL;				/* array MUST end with a null */
/*
** Now, initialize our toplevel widget and parse the command line
*/
 toplevel = XtInitialize(*aV, "XMail", Opts, XtNumber(Opts), &aC, aV);
 if (aC > 1) {
    (void) fprintf(stderr,"Usage: %s [-toolkitoptions] [-xmailoptions]\n",*aV);
    exit(-1);
    /*NOTREACHED*/
   }

 XtFree((char *)aV[0]);				/* we're done we these now */
 XtFree((char *)aV);
/*
** Next, add xmail specific as well as Athena text widget actions
*/
 XtAddActions(actionTable, XtNumber(actionTable));
 XtAddActions(_XawTextActionsTable, _XawTextActionsTableCount);

 XtGetApplicationResources(toplevel, &XMail, resrcs, XtNumber(resrcs), NULL, 0);

 XtOverrideTranslations(toplevel,
                     XtParseTranslationTable("<Message>WM_PROTOCOLS: Quit(q)"));

 initialize();				/* set window sizes based on font */
 CreateSubWindows(toplevel);		/* and build our widget hierarchy */
 XSync(XtDisplay(toplevel), False);	/* ensure everyone's on the same page */
 XtRealizeWidget(toplevel);		/* initialize widgets for our hints */
 SetHints();				/* tell WM about our icon and size */

 mailargc = mailoptions();		/* set up the mail pgm argv and argc */

 callMail(mailargv);			/* attempt to start Mail connection */

 SetCursor(WATCH);

 XtMainLoop();
 /* NOTREACHED */
} /* main */
